Create the Boa v0.17 release post (#94)
boa-dev/boa-dev.github.io: posts/2023-07-08-boa-release-17.md

---
layout: post
tags: post
title: "Boa release v0.17"
author: Boa Developers
---

## Summary

Boa v0.17 is now available! This is one of the biggest Boa releases since the project started, and after around 7 months
of development, we are very happy to present you the latest release of the Boa JavaScript engine. Boa makes it easy to
embed a JS engine in your projects, and you can even use it from WebAssembly. See the [about](/about) page for more info.

In this release, our conformance has grown from 74.53% to 78.74% in the official ECMAScript Test Suite (Test262). While
this might look as a small increase, we now pass **6,079** more tests than in the previous version. In any case, the big
changes in this release are not related to conformance, but to huge internal enhancements and new APIs that you will be
able to use.

You can check the full list of changes [here][changelog], and the full information on conformance [here][conformance].

Moreover, this big release was partly possible thanks to a grant by [Lit Protocol](https://litprotocol.com/). Lit is

> distributed key management for encryption, signing, and compute. Thanks to this grant, we were able to remunerate 2 team

Reviewer: I think we can omit explaining what lit protocol is, the hyperlink takes care of that for us.

Author: Done :)

members for their 20h/week work each during three and a half months. If you wish to sponsor Boa, you can do so by

> donating to out [open collective][collective]. You can also check [easy][easy_issues] or

Reviewer 2: *donating to **our**

Author: Fixed

// to pass it to the context.
let dyn_loader: &dyn ModuleLoader = loader;
let mut context = &mut Context::builder().module_loader(dyn_loader).build().unwrap();

let source = Source::from_bytes("1 + 3");

let module = Module::parse(source, None, context).unwrap();

// `main.mjs` or any of its imports could import `main.mjs` itself, so we
// insert it into the loader for good measure.
loader.insert(Path::new("main.mjs").to_path_buf(), module.clone());

// All modules use promises to signal completion of its lifecycle.
// The utility method `load_link_evaluate` calls `load`, then `link` and
// finally `evaluate`, returning an error if any call fails.
let promise = module.load_link_evaluate(context).unwrap();

// Important to push the job queue forward! Otherwise, the modules won't progress
// on their lifecycle.
context.run_jobs();

// All modules return `undefined` if they're successfully evaluated.
assert_eq!(promise.state().unwrap(), PromiseState::Fulfilled(JsValue::undefined()));
```

For a more extensive, descriptive example that uses a real directory, you can check out [**boa_examples**][mod_example].

[modules]: https://tc39.es/ecma262/#sec-modules

> [mod_loader]: https://boajs.dev/boa/doc/boa_engine/module/trait.ModuleLoader.html

Author: We should link to the 0.17 version of the docs
```suggestion
[mod_loader]: https://docs.rs/boa_engine/0.17.0/boa_engine/module/trait.ModuleLoader.html
```

> [mod_example]: https://github.com/boa-dev/boa/blob/main/boa_examples/src/bin/modules.rs

Author: ```suggestion
[mod_example]: https://github.com/boa-dev/boa/blob/v0.17/boa_examples/src/bin/modules.rs
```

### Spec Version Conformance

Something we get asked a lot is _"Do you support ES5 or ES6"_? or _"How far away are you from supporting ESX"_?
We're pleased to say we've updated our conformance board to show you how we're doing across ES versions.

> Just navigate to our [Test262 Dashboard](https://boajs.dev/boa/test262/), select "Test Results" on our main branch, and

Author: ```suggestion
Just navigate to our [Test262 Dashboard][conformance], select "Test Results" on our main branch, and
```

different from the traditional way of storing properties as a hashmap from property keys to values.

The shapes create a transition tree, where the transitions are property names and prototype changes starting from a root
shape (no properties, no prototype).

```javascript
let o = {} // Shape 1: prototype `Object.prototype` and properties: empty
o.a = 10   // Shape 2: prototype `Object.prototype` and properties: 'a'
o.b = 20   // Shape 3: prototype `Object.prototype` and properties: 'a', 'b'

let o2 = { a: 30; } // Shape 2: prototype `Object.prototype` and properties: 'a'
o2.d = 50           // Shape 4: prototype `Object.prototype` and properties: 'a', 'd' -- fork from shape 2
```

This separation of property keys and values allows for objects with the same property names to share the same shape,
which reduces memory consumption and unlocks the possibility for other optimizations such as inline caching.

Note: When creating objects with the same property keys, it's best to create them in the same order, this ensures that
the objects share the same shape.

For a more in depth explanation of how shared shapes work in boa see `shapes.md`

> [here](https://github.com/boa-dev/boa/blob/main/docs/shapes.md).

Author: ```suggestion
[here](https://github.com/boa-dev/boa/blob/v0.17/docs/shapes.md).
```

### Debug object

The `$boa` debug object has been implemented for convenient JavaScript debugging using Boa's CLI interface. If you want
to use it, you will need to run the Boa CLI / REPL with the `--debug-object` command line flag.

The `$boa` debug object is divided into modules, so that you can trigger the garbage collection with `$boa.gc.collect()`,
or get the bytecode of a function by running `$boa.function.bytecode(fn_name)`. You can also trace function invocations,
handle compiler optimizations, set runtime limits and inspect object shapes.

You can find all the documentation [here][debug_object].

> [debug_object]: https://github.com/boa-dev/boa/blob/main/docs/boa_object.md

Author: ```suggestion
[debug_object]: https://github.com/boa-dev/boa/blob/v0.17/docs/boa_object.md
```

        _ => Err(JsNativeError::typ().with_message("cannot convert value to an i16").into()),
    }
}

#[derive(Debug, TryFromJs)]
struct TestStruct {
    inner: bool,
    hello: String,
    // You can override the conversion of an attribute.
    #[boa(from_js_with = "lossy_conversion")]
    my_float: i16,
}

let js_str = r#"
let x = {
    inner: false,
    hello: "World",
    my_float: 2.9,
};
x;
"#;
let context = &mut Context::default();
let result = context.eval(Source::from_bytes(js_str))?;
let str = TestStruct::try_from_js(&result, context)?;

println!("{str:?}");
```

#### Source API

We have introduced a new `Source API` to Boa. The new API represents JavaScript stored from a path or `None` if it's
coming from a plain string.

This change improves the display of `boa_tester` to show the path of the tests being run. It also enables hyperlinks to
directly jump to the tested file from the VS terminal. This will further help with error displays and debugging in the
future.

```rust
use boa_engine::{Context, Source};

fn main() {
    let js_file_path = "./scripts/helloworld.js";

    match Source::from_filepath(Path::new(js_file_path)) {
    ...
```

> See Boa's [examples](https://github.com/boa-dev/boa/tree/main/boa_examples/src/bin) for more examples on how its used.

Author: ```suggestion
See Boa's [examples](https://github.com/boa-dev/boa/tree/v0.17/boa_examples/src/bin) for more examples on how its used.
```
